File: problems/0002-add-two-numbers/solution.py
from typing import Optional
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

def addDigits(l1: Optional[ListNode], l2: Optional[ListNode], carry: int) -> ListNode:
    assert(l1 is not None or l2 is not None)
    value: int = carry + (l1.val if l1 is not None else 0) + (l2.val if l2 is not None else 0)
    digitCarry: int = value // 10
    digitValue: int = value % 10
    hasL1Next: bool = l1 is not None and l1.next is not None
    hasL2Next: bool = l2 is not None and l2.next is not None
    if not hasL1Next and not hasL2Next:
        carryNode = ListNode(digitCarry) if digitCarry > 0 else None
        return ListNode(digitValue, carryNode)
    return ListNode(
        digitValue, 
        addDigits(
            l1.next if hasL1Next else None, 
            l2.next if hasL2Next else None, 
            digitCarry
        )
    )

class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if l1 is None and l2 is None:
            return None
        if l1 is None:
            return l2
        if l2 is None:
            return l1
        
        return addDigits(l1, l2, 0)

File: problems/0002-add-two-numbers/solution.ts
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
         this.next = (next===undefined ? null : next)
    }
}

function addDigits(l1: ListNode | null, l2: ListNode | null, carry: 0 | 1): ListNode {
    if (!l1 && !l2) {
        throw new Error('Invalid state, both digit nodes are null');
    }
    const value = carry + (l1?.val ?? 0) + (l2?.val ?? 0);
    const digitCarry = value > 9 ? 1 : 0;
    const digitValue = value % 10;
    if (!l1?.next && !l2?.next) {
        const carryNode = digitCarry ? new ListNode(digitCarry) : null;
        return new ListNode(digitValue, carryNode);
    }
    return new ListNode(digitValue, addDigits(l1?.next ?? null, l2?.next ?? null, digitCarry));
}

export function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    if (!l1 && !l2) {
        return null;
    }
    if (!l1) {
        return l2;
    }
    if (!l2) {
        return l1;
    }


    return addDigits(l1, l2, 0);
};